### PR TITLE
Tooltip Component

### PR DESCRIPTION
**Pull request by AngeCast**
Work for developing tooltip component.

Note: Logic in this component can get quite complex. Do let me know if more code comments are needed for clarification.

### CODE REVIEW

File: packages/stencil/src/components/datacom-tooltip/datacom-tooltip.tsx [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+    this.updateTooltipPosition(mainPosition);
+    let tooltipPositionData = this.getViewportPositionData(this.tooltipElement);
+
+    // Now determine if sub position should be start, end or center depending on where element is placed in viewport.
+    if (mainPosition === 'top' || mainPosition === 'bottom') {
+      // Vertical cardinal directions
+      if (tooltipPositionData.distanceFromLeft < 0 && tooltipPositionData.distanceFromRight >= 0) {
+        subPosition = 'start';
+      } else if (tooltipPositionData.distanceFromLeft >= 0 && tooltipPositionData.distanceFromRight < 0) {
+        subPosition = 'end';
```
Comment (sidra2022) [inline packages/stencil/src/components/datacom-tooltip/datacom-tooltip.tsx:153]: Please can you test this again, might be needed to change the subPosition value "start" to "end" and "end" to "start".

File: packages/stencil/src/components/datacom-tooltip/datacom-tooltip.tsx [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+  updateTooltipWidth(newWidth: number) {
+    if (newWidth === undefined) {
+      this.tooltipWrapperElement.style.removeProperty('width');
+    } else {
+      this.tooltipWrapperElement.style.width = `${newWidth}px`;
+    }
+  }
+
+  /**
+   * Returns data about element's visiblity in viewport and its distance from viewport's edges.
```
Comment (sidra2022) [inline packages/stencil/src/components/datacom-tooltip/datacom-tooltip.tsx:92]: Please can you check the visibility spelling?

File: packages/stencil/src/utils/utils.ts [hunk 3]
```diff
@@ -31,3 +31,17 @@ export function randomString(length = 10, prefix?: string): string {
     return (prefix || '') + random.map(n => n.toString(16)).join('');
   }
 }
+
+export const debounce = (func, wait: number) => {
+  let timeout: string | number | NodeJS.Timeout;
```
Comment (sidra2022) [inline packages/stencil/src/utils/utils.ts:36]: Please can you add comments for this file?

**Review by sidra2022 — COMMENTED** [review]
It is a complex component and few comments from me.
Also, I am just wondering why it has not "usage.md" file.

**Comment by AngeCast** [comment]
> It is a complex component and few comments from me according to my understanding. Also, I am just wondering why it has not "usage.md" file.

I didn't implement a usage.md file as we're now trying to move all the documentation to brandhub. That's the purpose of the dev documentation after all.

**Review by sidra2022 — APPROVED** [review]

**Pull request merged**